# Debug筆記 #001: ValueError - int()無法轉換文字

**只有「像數字」的字串才會變身成int()。**  
+ '123' ➜ 可以變身成int()  
+ 'hello' ➜ 不可以變身成int()，會報錯

In [14]:
#錯誤例子
x = 'Hello Bob'
y = int(nsv)

ValueError: invalid literal for int() with base 10: 'Hello Bob'

**錯誤原因說明**  
我嘗試用int()把一段不是數字的字串'Hello Bob'轉換成整數；  
但'Hello Bob'這種文字不能直接變成數字，Python會說「我看不懂這是什麼數字」，所以報錯。

In [7]:
#正確寫法
x = '123'
y = int(x)
print(y + 1)

124


# Debug筆記 #002: TypeError - 字串(String)不能做數學乘法

In [16]:
#錯誤例子: 練習input()時發生的錯誤
hours = input("請輸入你的工作時數 :")
rate = input("請輸入你的每小時薪水 :")
payment = hours * rate
print("Pay:", payment)

請輸入你的工作時數 : 35
請輸入你的每小時薪水 : 2.75


TypeError: can't multiply sequence by non-int of type 'str'

**錯誤原因說明**  
因為input()語法讀取進來的內容是「字串(String)」，  
但我們不能用「字串 * 字串」來算數學，Python會看不懂，所以報錯。

In [19]:
#正確寫法: 加上float()語法，將input()讀取到的字串(String)變成浮點數(Float)。
hours = float(input("請輸入你的工作時數 :"))
rate = float(input("請輸入你的每小時薪水 :"))
payment = hours * rate
print("Pay:", payment)

請輸入你的工作時數 : 35
請輸入你的每小時薪水 : 2.75


Pay: 96.25


# Debug筆記 #003: SyntaxError - print()字串str結合錯誤

In [20]:
#錯誤例子
money = 10
while money > 0:
    print("還有"money"元!")
    money = money - 2
print("沒錢啦! 小狐狸回家了!")

SyntaxError: invalid syntax. Perhaps you forgot a comma? (2802996702.py, line 4)

**錯誤原因說明**  
print("還有"money"元!")錯誤  
使用了多段字串str("還有"、"元")，中間卻插入了變數money。  
但是 **"還有"**、**money**、**"元"**，這三段被當成是 **沒有明確分隔的東西**，Python會不知道怎麼處理。  
不能直接把字串跟變數黏在一起使用。  

**以下共有四種正確寫法**  
(如果要去掉數字兩側的空格，推薦第四種寫法)

In [23]:
#正確寫法-1: 使用逗號(即Python的原始提示)
money = 10
print("還有", money, "元!")

還有 10 元!


In [26]:
#正確寫法-2: 使用逗號，但不想要空格，則使用spe=""強制去掉空格。
money = 10
print("還有", money, "元!", sep="")

還有10元!


In [27]:
#正確寫法-3: 使用字串加法，要加str()把變數money變成字串str。(同樣沒有空格)
money = 10
print("還有" + str(money) + "元!")

還有10元!


In [29]:
#正確寫法-4: 使用f-string (最乾淨最簡潔的方式)
money = 10
print(f"還有{money}元!")

還有10元!


關於f-string的應用請查閱Practice的**005_f-string_2025-06-06.ipnb**

# Debug筆記 #004: SyntaxError -  
# `指定賦值(assign) =` 和 `比較運算符號 ==` 混用

In [2]:
#錯誤例子
for i in range(1, 11):
    if i % 2 = 0:
        print(f"{i}是偶數")

SyntaxError: cannot assign to expression here. Maybe you meant '==' instead of '='? (1228459067.py, line 3)

**錯誤原因說明**  
不能在這裡使用「指定賦值(assign)」符號`=`。  
+ 條件判斷要用雙等號`==`，不是指定賦值符號`=`！

In [3]:
#正確寫法
for i in range(1,11):
    if i % 2 == 0:
        print(f"{i}是偶數")

2是偶數
4是偶數
6是偶數
8是偶數
10是偶數


**補充說明**  
+ 指定賦值(assign) → 使用語法 `=` → 把右邊的值給左邊的變數 (不能拿去做「判斷」)
+ 比較是否相等 → 使用語法 `==` → 判斷左右是否一樣 (if裡面該用這個)

# Debug筆記 #004: SyntaxError -  
# `else`語法錯誤 (不能加條件)

In [4]:
#錯誤例子
x = 5

if x < 10:
    print('Less than 10')
else x > 10:
    print('More than 10')

SyntaxError: expected ':' (3764762098.py, line 6)

**錯誤原因說明**  
`else`後面不可以加條件。

**正確語法規則**  
`if 條件:` → 如果符合這個條件就執行  
`elif 條件:` → 否則如果符合另一個條件就執行  
`else:` → 其他所有情況就執行 (**不能加條件**)  

In [8]:
#正確寫法-1
x = 5

if x < 10:
    print('Less than 10')
elif x > 10:
    print('More than 10')

Less than 10


In [7]:
#正確寫法-2 (如果只想判斷「不是小於10」的其他情況)
x = 20

if x < 10:
    print('Less than 10')
else:
    print('Not less than 10')

Not less than 10


# Debug筆記 #005: if-elif條件排序錯誤 - 
# 條件排序`要從最大條件往下排`

In [4]:
#錯誤例子
try:
    score = float(input("Enter Score: "))
    
    if score < 60:
        print("C")
    elif score >= 60:
        print("B")
    elif score >= 80:
        print("A")

except:
    print("Please Enter Numbers.")

Enter Score:  85


B


**錯誤原因說明**  
儘管程序並未報錯，但實際上運算只到第二階段`score >= 60`即停止，  
跳過了`score >= 80`的驗證步驟；所以得出的**結果無意義**。

**正確語法原則**  
如果條件有等級分類(例如成績、價錢、年齡階段)，**要從最大條件往下排**。  
+ if-elif的條件順序非常重要，一旦某個條件成立，Python就不會再繼續看後面的elif。
+ 寫等級條件時，要從`範圍最大的條件往下排`，才會判斷正確。

In [2]:
#正確寫法
try:
    score = float(input("Enter Score: "))

    if score >= 80:
        print("A")
    elif score >= 60:
        print("B")
    elif score < 60:
        print("C")

except:
    print("Please Enter Numbers.")

Enter Score:  85


A


# Debug筆記 #006: TypeError (類型錯誤) - 
# `f-string`中不能混用str + int  

In [5]:
#錯誤例子
x = '123'
print(f"{x + 1}")

TypeError: can only concatenate str (not "int") to str

**錯誤原因說明**  
`x`是字串(str)型別，但是1是整數(int)型別，  
在`f-string`裡，`{x + 1}`實際是執行`str + int`的運算，這會導致類型錯誤(TypeError)。  
`+`對於字串代表「拼接」，但不能讓字串跟整數直接拼接。

In [7]:
# 正確寫法 (如果是要數字計算)
x = '123'
y = int(x)
print(f"{y + 1}")

124


In [12]:
# 正確寫法 (如果是需要字串)
x = '123'
print(f"{x + '1'}")

# 用 '...' 將數字1轉成字串1

1231


**正確語法原則**  
**f-string只是顯示格式**，內部`{}`仍然執行標準Python運算。  
使用`+`時，請確認兩邊型別一致:
+ `str + str` (正確)
+ `int + int` (正確)
+ `str + int` (錯誤，需轉型)

#### f-string中`{}`裡的行為對照表：
`{123 + 1}` → `結果: 124` (數字加法)  
`{'123' + '1'}` → `結果: '1231'` (字串拼接)  
`{'123' + 1}` → `錯誤結果: TypeError` (`str + int`不合法)  
`{int('123'} + 1}` → `結果: 124` (將字串轉為整數，再相加)  
`{str(123) + '1'}` → `結果: ‘1231` (將整數轉為字串，再拼接)  

# Debug筆記 #007: 函式未回傳值導致`None`
## (`return`語法 v.s. `print()`語法)

+ 參考語法筆記: `010_Function_Definition_2025-06-09`

In [2]:
# 錯誤例子 (進階)

def greet_user(name, age):
    if age < 18:
        print(f"{name}，你還是學生吧?")
    else:
        print(f"{name}，歡迎加入社會!")

print(greet_user("奶油酥餅", 17))

奶油酥餅，你還是學生吧?
None


**錯誤現象**  
執行`print(greet_user(...))`時，畫面出現正確訊息後，又多印出一行`None`。

**錯誤原因說明**  
在函式`greet_user()`裡使用了`print()`而非`return`，  
導致函式沒有回傳任何值，預設傳回`None`，  
這個`None`被外層的`print()`顯示出來。

**正確寫法**  
改用`return`回傳字串給外層`print()`:  

In [4]:
#正確寫法

def greet_user(name, age):
    if age < 18:
        return f"{name}, 你還是學生吧?"
    else:
        return f"{name}, 歡迎加入社會!"

print(greet_user("奶油酥餅", 17))

奶油酥餅, 你還是學生吧?


**語法概念重點說明**  
`print()` : 直接在函示內輸入內容，但不會回傳結果  
`return` : 回傳資料給呼叫者，可由外層`print()`印出  
`函式回傳` : 如果函式沒有return，Python預設回傳`None`  